File: src/main/java/org/vanautrui/languages/compiler/parsing/astnodes/typenodes/SubroutineTypeNode.java
package org.vanautrui.languages.compiler.parsing.astnodes.typenodes;

import org.vanautrui.languages.compiler.lexing.tokens.ArrowToken;
import org.vanautrui.languages.compiler.lexing.tokens.SymbolToken;
import org.vanautrui.languages.compiler.lexing.utils.TokenList;
import org.vanautrui.languages.compiler.parsing.IASTNode;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

class SubroutineTypeNode implements ITypeNode {

  public final ITypeNode returnType;

  public final boolean has_side_effects;

  public List<TypeNode> argumentTypes=new ArrayList<>();

  public SubroutineTypeNode(TokenList tokens)throws Exception{

    TokenList copy = tokens.copy();

    copy.expectAndConsumeOtherWiseThrowException(new SymbolToken("("));

    boolean sucess_argument_types=true;
    try {
      this.argumentTypes.add(new TypeNode(copy));
    }catch (Exception e){
      sucess_argument_types=false;
    }
    while (sucess_argument_types){
      try {
        TokenList copy2 = copy.copy();

        copy2.expectAndConsumeOtherWiseThrowException(new SymbolToken(","));
        this.argumentTypes.add(new TypeNode(copy2));

        copy.set(copy2);
      }catch (Exception e){
        sucess_argument_types=false;
      }
    }

    copy.expectAndConsumeOtherWiseThrowException(new SymbolToken(")"));

    if(copy.head() instanceof ArrowToken){
      ArrowToken arrow = (ArrowToken) copy.head();
      this.has_side_effects=!arrow.is_functional;
      copy.consume(1);
    }else{
      throw new Exception("expected an arrow token here");
    }

    if(copy.head().tokenEquals(new SymbolToken("("))){
      //return type is subroutine
      copy.expectAndConsumeOtherWiseThrowException(new SymbolToken("("));
      this.returnType = new SubroutineTypeNode(copy);
      copy.expectAndConsumeOtherWiseThrowException(new SymbolToken(")"));
    }else {
      this.returnType = new SimpleTypeNode(copy);
    }

    tokens.set(copy);
  }

  public String getTypeName(){
    return this.toSourceCode();
  }

  @Override
  public String toSourceCode() {
    if(this.returnType instanceof SubroutineTypeNode){
      return
              "("
                      + argumentTypes
                      .stream()
                      .map(IASTNode::toSourceCode)
                      .collect(Collectors.joining(","))
                      + ")" + ((this.has_side_effects) ? "~>" : "->") + "("+returnType.toSourceCode()+")";
    }else {
      return
              "("
                      + argumentTypes
                      .stream()
                      .map(IASTNode::toSourceCode)
                      .collect(Collectors.joining(","))
                      + ")" + ((this.has_side_effects) ? "~>" : "->") + returnType.toSourceCode();
    }
  }
}

File: dragon-parser/src/main/compiler/parsing/astnodes/typenodes/basic_and_wrapped/SubroutineTypeNode.cpp

//standard headers
#include <vector>
#include <set>

//project headers
#include "SubroutineTypeNode.hpp"

public final class SubroutineTypeNode implements IBasicAndWrappedTypeNode  {

	public final TypeNode returnType;

	public final boolean hasSideEffects;

	public final List<TypeNode> argumentTypes = new ArrayList<>();

	public SubroutineTypeNode(final TypeNode return_type, final boolean hasSideEffects) {
		this.returnType = return_type;
		this.hasSideEffects = hasSideEffects;
	}

	public SubroutineTypeNode(final TokenList tokens) throws Exception {

		final TokenList copy = tokens.copy();

		copy.expectAndConsumeOtherWiseThrowException(new LParensToken());

		boolean sucess_argument_types = true;
		try {
			this.argumentTypes.add(new TypeNode(copy));
		} catch (Exception e) {
			sucess_argument_types = false;
		}
		while (sucess_argument_types) {
			try {
				TokenList copy2 = copy.copy();

				copy2.expectAndConsumeOtherWiseThrowException(new CommaToken());
				this.argumentTypes.add(new TypeNode(copy2));

				copy.set(copy2);
			} catch (Exception e) {
				sucess_argument_types = false;
			}
		}

		copy.expectAndConsumeOtherWiseThrowException(new RParensToken());

		if (copy.head() instanceof ArrowToken) {
			ArrowToken arrow = (ArrowToken) copy.head();
			this.hasSideEffects = !arrow.is_functional;
			copy.consume(1);
		} else {
			throw new Exception("expected an arrow token here");
		}

		this.returnType = new TypeNode(copy);

		tokens.set(copy);
	}

}
